.travis.yml

```diff
@@ -0,0 +1,106 @@
+# Language and cache settings
+language: python
+python: 2.7
+cache:
+  pip: true
+  directories:
+    - $HOME/.cache/apt
+
+# Common install step
+install:
+  # Make sure pipefail
+  - set -o pipefail
+  # Set up apt to cache
+  - mkdir -p $HOME/.cache/apt/partial
+  - sudo rm -rf /var/cache/apt/archives
+  - sudo ln -s $HOME/.cache/apt /var/cache/apt/archives
+  # Set up ppa to make sure arm-none-eabi-gcc is the correct version
+  - sudo add-apt-repository -y ppa:team-gcc-arm-embedded/ppa
+  # Fix for "The following signatures were invalid: KEYEXPIRED 1515625755" failed". See https://github.com/travis-ci/travis-ci/issues/9037
+  - sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv 0C49F3730359A14518585931BC711F9BA15703C6
+  # Loop until update succeeds (timeouts can occur)
+  - travis_retry $(! sudo apt-get update 2>&1 | grep Failed)
+  # Finally install gcc-arm-embedded
+  - sudo apt-get install gcc-arm-embedded
+  # Get mbed-os
+  - git clone https://github.com/armmbed/mbed-os.git
+  # Install python dependencies
+  - pip install -r mbed-os/requirements.txt
+
+# CI matrix
+jobs:
+  include:
+    # native testing
+    - stage: test
+      env:
+        - NAME=code-snippets
+      script:
+        # Run script to validate code snippets
+        - ./check_tools/find_bad_code_snippets.sh
+        # Grep to find remaining TODOs
+        - |
+            TODO_COUNT=0
```

Comment:
What this TODO means here? Is it "NOT_COMPILE" like ? I would expect not having any todo on the documentation we publish but rather have the code that should compile if I copy-paste it or code that is to illustrate - stated there and defined as "not ready for compilation, needs edits"

Reply:
Ah I didn't explain it that well did I. @SenRamakri asked to present this next demo day, so I was going to go into more explanation there.

Basically there are 3 types of code snippets now:
- ` ``` cpp `
  By default code snippets should compile. You can opt-out with the other tags, but if you're just being lazy @iriark01 or @AnotherButler may get mad.

- ` ``` cpp NOCI `
  Code snippets that are not expected to compile. This may be snippets with only a single statement or an example of broken code (also makes @iriark01+@AnotherButler mad). If the snippets more than a handful of statements it should be wrapped in a function so that it can be compiled.

- ` ``` cpp TODO `
  Code snippets that should compile but don't. This was added because we currently have a lot of examples that should compile but don't. I tried to fix them where I could, but when I referenced the mbed docs the examples were broken there too!

  From talking with @AnotherButler, I believe the current plan is to take code snippets with TODO and create tasks for each of them for the owner of each technology to clean up. A lot of them should be simple (I think the USB examples just need the correct headers?).

You can find the TODO list in Travis log, and the bot should be responding with the number of TODOs so that the doc writers can see any changes (similar to the astyle reporting). But that's currently waiting on adding @mbed-bot to the users with push access. I was going to copy and paste that TODO list here once it's ready to go.

Reply:
> we wont run any validation

That's what it's saying isn't it? What other validation are we running. IMO WONT_FIX isn't right since it shouldn't be broken.

---

Yep, to be clear the goal is to eventually remove all TODOs.

Maintaining a separate branch for TODO cleanup is not going to be tractable. This affects the entire handbook and the handbook moves too fast. The PR already has conflicts despite being up for only a day.

This is the same strategy @0xc0170 is applying with the astyle fixes in mbed-os, and it's the best approach I can think of without disrupted everyone who is currently working on handbook changes.

Reply:
@geky - We don't have this script to be gating now although we should start running(let it fail for now) so everyone knows we have failing snippets. Actually, if we let the script pass with all TODOs now it becomes slower to fix them.

Reply:
This has come up before, but the problem is that all PRs will have red Xs for thing the PR owner can't fix. This leads to users simply ignoring the CI results.

Reply:
I personally prefer the TODO method. Like @geky said if we push this in without that feature and let the multiple dozens of broken code just fail people will likely tend to disregard the CI results completely. Having a counter for remaining TODOs and having action items in Jira/Github to fix each TODO with a corresponding owner makes more sense to me.